### PR TITLE
ADBDEV-5499: contrib: enable using tablefunc for CTAS/DML queries

### PR DESCRIPTION
**Pull request by silent-observer**
When user executes CTAS or DML query with use of functions from `tablefunc`
extension (for ex. `insert into table select * from crosstab(...)`),
next error occurs:
`function cannot execute on a QE slice because it accesses relation`.
In case of these operations `tablefunc` functions needs to be dispatched,
so `EXECUTE ON INITPLAN` option was added to them
(this solution based on next commit: b387ef3).

* Added `contrib/tablefunc/{Makefile, data}` (from gpdb upstream).
* Enabled build of `tablefunc` extension at:
    - `contrib/Makefile`
    - `GNUmakefile.in`
* Added new version of `tablefunc` extension at `contrib/tablefunc`:
    - added `tablefunc--1.1.sql` install script for new version of
    extension, which supports CTAS/DML queries
    - changed default extension version at `tablefunc.control` to 1.1
    - added `tablefunc--1.0--1.1.sql` update script
* Added tests for CTAS operation at:
    - `contrib/tablefunc/{expected, sql, data}`

Cherry-picked-from: 3a961c0e1243b315c22440634250ef142da8edf4
to solve conflicts with 848d5e6

Changes from original commit:
- Removed tablefunc--1.1.sql since it is no longer necessary for newer
versions of Postgres.
- Fix Makefile to account for that.
- Removed ON INITPLAN from normal_rand since it is unnecessary.
- Added missing tests.

(cherry picked from commit f6bcded1e716584824034234c7762e3f18d5956a)

---

Note: do not squash to preserve authorship

### CODE REVIEW

File: contrib/tablefunc/tablefunc--1.1.sql [hunk 1]
```diff
@@ -0,0 +1,88 @@
+/* contrib/tablefunc/tablefunc--1.1.sql */
```
Comment (whitehawk) [inline contrib/tablefunc/tablefunc--1.1.sql]: Do we really need this file? According to [Postgres 12 documentation](https://www.postgresql.org/docs/12/extend-extensions.html): 
> if you have released a foo extension in versions 1.0, 1.1, and 1.2, there should be update scripts foo--1.0--1.1.sql and foo--1.1--1.2.sql. Before PostgreSQL 10, it was necessary to also create new script files foo--1.1.sql and foo--1.2.sql that directly build the newer extension versions, or else the newer versions could not be installed directly, only by installing 1.0 and then updating. That was tedious and duplicative, but now it's unnecessary

so I guess having 'tablefunc--1.0--1.1.sql' should be enough...

**Comment by whitehawk** [comment]
Please add a note to the PR description that the commit should be rebased to preserve authorship.

File: contrib/tablefunc/tablefunc--1.1.sql (hunk 1 above)
Comment (silent-observer) [reply contrib/tablefunc/tablefunc--1.1.sql]: Fixed

**Review by whitehawk — DISMISSED** [review]

File: contrib/tablefunc/tablefunc--1.0--1.1.sql [hunk 2]
```diff
@@ -0,0 +1,16 @@
+/* contrib/tablefunc/tablefunc--1.0--1.1.sql */
+
+-- complain if script is sourced in psql, rather than via ALTER EXTENSION
+\echo Use "ALTER EXTENSION tablefunc UPDATE TO '1.1'" to load this file. \quit
+
+ALTER function normal_rand(integer,double precision,double precision) EXECUTE ON INITPLAN;
```
Comment (hilltracer) [inline contrib/tablefunc/tablefunc--1.0--1.1.sql]: Do we really need to add normal_rand on initplan? Can you show any examples with reproducing the problem. Seems like it works fine in old code. E.g.
```sql
create table t_avg as select * from normal_rand(5, 250, 0);
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 3]
```diff
@@ -204,3 +204,95 @@ INSERT INTO connectby_int VALUES(111,11);
 INSERT INTO connectby_int VALUES(1,111);
 -- this should not fail due to recursion detection
 SELECT * FROM connectby('connectby_int', 'keyid', 'parent_keyid', '11', 0, '-') AS t(keyid int, parent_keyid int, level int, branch text);
+
+
+-- test for CTAS and DML operations
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql:208]: Looks like test covers only 3 functions `crosstab(text,text)`, `crosstab(text)`, `connectby(text,text,text,text,integer,text)`.
Is it ok that other functions are not checked?
e.g.
```sql
create table ct2 as SELECT * FROM crosstab2('SELECT rowid, attribute, val FROM ct where rowclass = ''group1'' and (attribute = ''att2'' or attribute = ''att3'') ORDER BY 1,2;');
# ERROR:  function cannot execute on a QE slice because it accesses relation "public.ct"  (seg0 slice1 172.18.0.3:7002 pid=50171)
```

Comment (silent-observer) [reply contrib/tablefunc/sql/tablefunc.sql:208]: Do we need to? The issue is fairly straightforward and the same across all of them, I think there is no benefit for checking all of them separately

Comment (hilltracer) [reply contrib/tablefunc/sql/tablefunc.sql:208]: I'm not sure, for my opinion tests should cover all threads in the code. But in current case this requires writing a lot of extra code. @whitehawk what do you think?

**Comment by hilltracer** [comment]
Please update PR description to reflect the latest change.

File: contrib/tablefunc/sql/tablefunc.sql (hunk 3 above)
Comment (whitehawk) [reply contrib/tablefunc/sql/tablefunc.sql:208]: I think it is better to add more test cases to avoid any possible breakages in future.

Comment (silent-observer) [reply contrib/tablefunc/sql/tablefunc.sql:208]: Done

File: contrib/tablefunc/sql/tablefunc.sql [hunk 4]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+-- from normal_dist
+CREATE TABLE t_normal(normal_rand float8);
+
+INSERT INTO t_normal SELECT * FROM normal_rand(5, 250, 0);
+
+SELECT * FROM t_normal;
+
+CREATE TABLE t_normal_2 AS SELECT * FROM normal_rand(5, 250, 0);
+
+SELECT * FROM t_normal_2;
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql]: Missing empty line at the end of the file.

File: contrib/tablefunc/sql/tablefunc.sql [hunk 5]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
 INSERT INTO connectby_int VALUES(1,111);
 -- this should not fail due to recursion detection
 SELECT * FROM connectby('connectby_int', 'keyid', 'parent_keyid', '11', 0, '-') AS t(keyid int, parent_keyid int, level int, branch text);
+
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql:207]: Extra empty line

File: contrib/tablefunc/sql/tablefunc.sql [hunk 6]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+SELECT * FROM connby3;
+
+CREATE TABLE connby3_2 AS SELECT * FROM connectby(
+  'connectby_text', 'keyid', 'parent_keyid', 'pos', 'row2', 0, '~'
+) AS t(keyid text, parent_keyid text, level int, branch text, pos int) ORDER BY t.pos;
+
+SELECT * FROM connby3_2;
+
+-- Withbranch, without orderly
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql]:
```suggestion
-- from connectby() with branch, without orderby
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 7]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+SELECT * FROM ct6_2;
+
+CREATE TABLE ct7 (
+    row_name   text,
+    category_1 text,
+    category_2 text,
+    category_3 text,
+    category_4 text
+) DISTRIBUTED BY (row_name);
+INSERT INTO ct7
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql:349]:
```suggestion

INSERT INTO ct7
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 8]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+SELECT * FROM ct2_2;
+
+-- from crosstab(text)
+CREATE TABLE ct3 (
+    row_name varchar(32),
+    cat1     varchar(32),
+    cat2     varchar(32),
+    cat3     varchar(32)
+) DISTRIBUTED BY (row_name);
+INSERT INTO ct3
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql:258]:
```suggestion

INSERT INTO ct3
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 9]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+)
+DISTRIBUTED BY (row_name);
+\copy ct4 from 'data/ctas.data'
+
+CREATE TABLE ct5 (
+    row_name   text,
+    category_1 text,
+    category_2 text
+) DISTRIBUTED BY (row_name);
+INSERT INTO ct5
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql:295]:
```suggestion

INSERT INTO ct5
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 10]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+   order by 1,2') AS ct ;
+SELECT * FROM ct5_2;
+
+CREATE TABLE ct6 (
+    row_name   text,
+    category_1 text,
+    category_2 text,
+    category_3 text
+) DISTRIBUTED BY (row_name);
+INSERT INTO ct6
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql:322]:
```suggestion

INSERT INTO ct6
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 11]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+FROM crosstab(
+  'select row_name, attr, value
+   from ct1
+   where attr = ''att2'' or attr = ''att3''
+   order by 1,2')
+AS ct(row_name varchar, att1 varchar, att2 varchar, att3 varchar);
+
+SELECT * FROM ct3_2;
+
+-- from crosstab3(text)
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql]:
```suggestion
-- from crosstab2(text)
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 12]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+
+CREATE TABLE ct5_2 AS SELECT *
+FROM crosstab2(
+  'select row_name, attr, value
+   from ct4
+   where attr = ''att2'' or attr = ''att3''
+   order by 1,2') AS ct ;
+SELECT * FROM ct5_2;
+
+CREATE TABLE ct6 (
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql:315]:
```suggestion
-- from crosstab3(text)
CREATE TABLE ct6 (
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 13]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+INSERT INTO ct7
+SELECT *
+FROM crosstab3(
+  'select row_name, attr, value
+   from ct4
+   where attr = ''att2'' or attr = ''att3''
+   order by 1,2');
+SELECT * FROM ct7;
+
+CREATE TABLE ct7_2 AS SELECT *
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql:350]:
```suggestion
-- from crosstab4(text)
CREATE TABLE ct7_2 AS SELECT *
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 14]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+
+-- from crosstab3(text)
+CREATE TABLE ct4(
+    row_name text,
+    extra_col text,
+    attr text,
+    value text
+)
+DISTRIBUTED BY (row_name);
+\copy ct4 from 'data/ctas.data'
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql:287]: Why did you change `varchar(32)` to `text` in the `crosstab2..4` tests? Can we save previous tests style, use ct1 in these tests and remove ct4?

File: contrib/tablefunc/sql/tablefunc.sql [hunk 15]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+   order by 1,2');
+SELECT * FROM ct6;
+
+CREATE TABLE ct6_2 AS SELECT *
+FROM crosstab3(
+  'select row_name, attr, value
+   from ct4
+   where attr = ''att2'' or attr = ''att3''
+   order by 1,2') AS ct ;
+SELECT * FROM ct6_2;
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql:338]:
```suggestion

SELECT * FROM ct6_2;
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 16]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+    category_3 text
+) DISTRIBUTED BY (row_name);
+INSERT INTO ct6
+SELECT *
+FROM crosstab3(
+  'select row_name, attr, value
+   from ct4
+   where attr = ''att2'' or attr = ''att3''
+   order by 1,2');
+SELECT * FROM ct6;
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql:329]:
```suggestion

SELECT * FROM ct6;
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 17]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+
+SELECT * FROM connby;
+
+CREATE TABLE connby_2 AS SELECT * FROM connectby(
+  'connectby_text', 'keyid', 'parent_keyid', 'pos', 'row2', 0
+) AS t(keyid text, parent_keyid text, level int, pos int) ORDER BY t.pos;
+
+SELECT * FROM connby_2;
+
+-- Without branch, without orderly
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql]:
```suggestion
-- from connectby() without branch, without orderby
```

File: contrib/tablefunc/sql/tablefunc.sql [hunk 18]
```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
...
+
+SELECT * FROM connby2;
+
+CREATE TABLE connby2_2 AS SELECT * FROM connectby(
+  'connectby_text', 'keyid', 'parent_keyid', 'row2', 0
+) AS t(keyid text, parent_keyid text, level int);
+
+SELECT * FROM connby2_2;
+
+-- With branch, with orderly
```
Comment (hilltracer) [inline contrib/tablefunc/sql/tablefunc.sql]:
```suggestion
-- from connectby() with branch, with orderby
```

File: contrib/tablefunc/sql/tablefunc.sql (hunk 14 above)
Comment (silent-observer) [reply contrib/tablefunc/sql/tablefunc.sql:287]: No, these functions return custom types `tablefunc_crosstab_2..4` which have `text`, so the output types have to match. Meanwhile the plain `crosstab` is generic so it returns `record` and type conversion happens inside `AS ct` clause (which isn't supported for custom types). Changing types to `text` is the easiest solution

**Review by hilltracer — APPROVED** [review]

File: contrib/tablefunc/tablefunc--1.0--1.1.sql [hunk 19]
```diff
@@ -0,0 +1,15 @@
+/* contrib/tablefunc/tablefunc--1.0--1.1.sql */
+
+-- complain if script is sourced in psql, rather than via ALTER EXTENSION
+\echo Use "ALTER EXTENSION tablefunc UPDATE TO '1.1'" to load this file. \quit
+
+ALTER function crosstab(text) EXECUTE ON INITPLAN;
+ALTER function crosstab2(text) EXECUTE ON INITPLAN;
+ALTER function crosstab3(text) EXECUTE ON INITPLAN;
+ALTER function crosstab4(text) EXECUTE ON INITPLAN;
+ALTER function crosstab(text,integer) EXECUTE ON INITPLAN;
```
Comment (whitehawk) [inline contrib/tablefunc/tablefunc--1.0--1.1.sql:10]: Isn't this one (`crosstab(text,integer)`) missing in the test?

Comment (silent-observer) [reply contrib/tablefunc/tablefunc--1.0--1.1.sql:10]: That one is declared obsolete and isn't tested even in already existing part of the test. I think we don't need to test it

**Review by whitehawk — APPROVED** [review]
